lib/std/mem.zig

```diff
@@ -57,23 +86,64 @@ pub const Allocator = struct {
         /// Guaranteed to be a power of 2.
         /// Returned slice's pointer must have this alignment.
         new_alignment: u29,
-    ) Error![]u8,
+    ) Error![]u8 {
+        if (old_mem.len == 0)
+            return self.allocMem(new_byte_count, new_alignment);
+        if (new_byte_count == 0) {
+            self.freeMem(old_mem);
+            return old_mem[0..0];
+        }
+        if (isAligned(@ptrToInt(old_mem.ptr), new_alignment)) {
+            blk: {
+                self.resizeMem(old_mem, new_byte_count) catch |e| switch (e) {
+                    error.OutOfMemory => break :blk,
+                };
+                return old_mem.ptr[0..new_byte_count];
+            }
+        }
+        if (new_byte_count <= old_mem.len and new_alignment <= old_alignment) {
+            return error.OutOfMemory;
+        }
+        return self.moveMem(old_mem, new_byte_count, new_alignment);
+    }
 
-    /// This function deallocates memory. It must succeed.
-    shrinkFn: fn (
+    /// Move the given memory to a new location in the given allocator to accomodate a new
+    /// size and alignment.
+    fn moveMem(self: *Allocator, old_mem: []u8, new_len: usize, new_alignment: u29) Error![]u8 {
+        assert(old_mem.len > 0);
+        assert(new_len > 0);
+        const new_mem = try self.allocMem(new_len, new_alignment);
+        @memcpy(new_mem.ptr, old_mem.ptr, std.math.min(new_len, old_mem.len));
+        self.freeMem(old_mem);
+        return new_mem;
+    }
+
+    /// This function attempts to deallocates memory. It must succeed.
+    fn shrinkMem(
         self: *Allocator,
         /// Guaranteed to be the same as what was returned from most recent call to
-        /// `reallocFn` or `shrinkFn`.
+        /// `reallocFn` or `shrinkMem`.
         old_mem: []u8,
         /// Guaranteed to be the same as what was returned from most recent call to
-        /// `reallocFn` or `shrinkFn`.
+        /// `reallocFn` or `shrinkMem`.
         old_alignment: u29,
         /// Guaranteed to be less than or equal to `old_mem.len`.
         new_byte_count: usize,
         /// If `new_byte_count == 0` then this is `undefined`, otherwise:
         /// Guaranteed to be less than or equal to `old_alignment`.
         new_alignment: u29,
-    ) []u8,
+    ) []u8 {
+        assert(new_byte_count <= old_mem.len);
+        assert(new_alignment <= old_alignment);
+        if (new_byte_count == 0) {
+            self.freeMem(old_mem);
+        } else if (isAligned(@ptrToInt(old_mem.ptr), new_alignment)) {
+            self.resizeMem(old_mem, new_byte_count) catch |e| switch (e) {
+                error.OutOfMemory => {}, // ignore the error, we made a best effort to shrink
+            };
```

Comment:
Should we also try to `moveMem`? Slab allocators tend to never shrink in-place so this could be wasteful.

Reply:
It looks like zig tests are expecting `shrink` not to allocate new memory.

Reply:
Ah okay, it might be a specific test case for FixedBufferAllocator. With optional shrinkFn, we can possibly do this more intelligently.